x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.test.tsx

```diff
@@ -0,0 +1,311 @@
+  encodeRisonUrlState: jest.fn().mockImplementation((value) => value),
+}));
+jest.mock('./rules_table_context');
+
+describe('useSyncRulesTableSavedState', () => {
+  const defaultState = {
+    filterOptions: DEFAULT_FILTER_OPTIONS,
+    sortingOptions: DEFAULT_SORTING_OPTIONS,
+    pagination: {
+      page: DEFAULT_PAGE,
+      perPage: DEFAULT_RULES_PER_PAGE,
+      total: 100,
+    },
+  };
+  const expectStateToSyncWithUrl = (
+    rulesTableState: Partial<RulesTableState>,
+    expectedUrlState: RulesTableUrlSavedState
+  ) => {
+    (useRulesTableContext as jest.Mock).mockReturnValue({
+      state: rulesTableState,
+    });
+
+    renderHook(() => useSyncRulesTableSavedState());
+
+    expect(replaceUrlParams).toHaveBeenCalledWith([
+      {
+        key: URL_PARAM_KEY.rulesTable,
+        value: expectedUrlState,
+      },
+    ]);
+  };
+  const expectStateToSyncWithStorage = (
+    rulesTableState: Partial<RulesTableState>,
+    expectedStorageState: RulesTableStorageSavedState
+  ) => {
+    (useRulesTableContext as jest.Mock).mockReturnValue({
+      state: rulesTableState,
+    });
+
+    renderHook(() => useSyncRulesTableSavedState());
+
+    expect(setStorage).toHaveBeenCalledWith(RULES_TABLE_STATE_STORAGE_KEY, expectedStorageState);
+  };
+
+  let replaceUrlParams: jest.Mock;
+  let setStorage: jest.Mock;
+  let removeStorage: jest.Mock;
+
+  beforeEach(() => {
+    replaceUrlParams = jest.fn();
+    setStorage = jest.fn();
+    removeStorage = jest.fn();
+
+    (useReplaceUrlParams as jest.Mock).mockReturnValue(replaceUrlParams);
+    (useKibana as jest.Mock).mockReturnValue({
+      services: { sessionStorage: { set: setStorage, remove: removeStorage } },
+    });
+  });
+
+  it('clears the default state when there is nothing to sync', () => {
+    (useRulesTableContext as jest.Mock).mockReturnValue({ state: defaultState });
+
+    renderHook(() => useSyncRulesTableSavedState());
+
+    expect(replaceUrlParams).toHaveBeenCalledWith([{ key: URL_PARAM_KEY.rulesTable, value: null }]);
+    expect(setStorage).not.toHaveBeenCalled();
+    expect(removeStorage).toHaveBeenCalledWith(RULES_TABLE_STATE_STORAGE_KEY);
+  });
+
+  it('syncs both the url and the storage', () => {
+    const state = {
+      filterOptions: {
+        filter: 'test',
+        tags: ['test'],
+        showCustomRules: true,
+        showElasticRules: false,
+      },
+      sortingOptions: {
+        field: 'name',
+        order: 'asc',
+      },
+      pagination: {
+        page: 3,
+        perPage: 10,
+        total: 100,
+      },
+    };
+    const expectedUrlState: RulesTableUrlSavedState = {
+      searchTerm: 'test',
+      source: RuleSource.Custom,
+      tags: ['test'],
+      field: 'name',
+      direction: 'asc',
+      page: 3,
+      perPage: 10,
+    };
+    const expectedStorageState: RulesTableStorageSavedState = omit(expectedUrlState, 'page');
+
+    (useRulesTableContext as jest.Mock).mockReturnValue({
+      state,
+    });
```

Comment:
Can the `useRulesTableContextMock` be used here as well? The idea is to make mocks strongly typed. If context types change, we should be able to see all the affected files easily.

Example of how the mock is used in other places:

https://github.com/elastic/kibana/blob/2efe77b5d847797c6d4763d540a117044fd69a81/x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table_utility_bar.test.tsx#L118-L122

Reply:
Right, just pushed it.